ports/geckolib/context.rs

```diff
@@ -0,0 +1,67 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+use selector_impl::{Animation, GeckoSelectorImpl, SharedStyleContext};
+use std::cell::RefCell;
+use std::rc::Rc;
+use std::sync::Mutex;
+use std::sync::mpsc::Sender;
+use style::context::{LocalStyleContext, StyleContext};
+use style::matching::{ApplicableDeclarationsCache, StyleSharingCandidateCache};
+
+thread_local!(static LOCAL_CONTEXT_KEY:
+                RefCell<Option<Rc<LocalStyleContext<GeckoSelectorImpl>>>> = RefCell::new(None));
+
+// Keep this implementation in sync with the one in components/layout/context.rs.
+fn create_or_get_local_context(shared: &SharedStyleData)
+                               -> Rc<LocalStyleContext<GeckoSelectorImpl>> {
+    LOCAL_CONTEXT_KEY.with(|r| {
+        let mut r = r.borrow_mut();
+        if let Some(context) = r.clone() {
+            if shared.shared_style_context.screen_size_changed {
+                context.applicable_declarations_cache.borrow_mut().evict_all();
+            }
+            context
+        } else {
+            let new_animations_sender = shared.new_animations_sender.lock().unwrap().clone();
+            let context = Rc::new(LocalStyleContext {
+                applicable_declarations_cache: RefCell::new(ApplicableDeclarationsCache::new()),
+                style_sharing_candidate_cache: RefCell::new(StyleSharingCandidateCache::new()),
+                new_animations_sender: new_animations_sender,
+            });
+            *r = Some(context.clone());
+            context
+        }
+    })
+}
+
+/// A shared style context, plus data that we need to create a local one.
+pub struct SharedStyleData {
+    pub shared_style_context: SharedStyleContext,
+    pub new_animations_sender: Mutex<Sender<Animation>>,
+}
+
+pub struct StandaloneStyleContext<'a> {
+    pub shared_data: &'a SharedStyleData,
```

Comment:
More generally though, I'm not super convinced of the usefulness here. Wouldn't it be better to just leave the Mutexed Sender on the SharedStyleData, and make a local un-Mutexed clone for the local style context? That would avoid the need to introduce a new data structure here with someone-ambiguous meaning, and also avoid the need to add it to SharedLayoutContext above.